src/python/pants/java/junit/junit_xml_parser.py

```diff
@@ -68,6 +69,44 @@ def empty(self):
     """
     return len(self._test_to_target) == 0
 
+  def match_test_spec(self, possible_test_specs):
+    """
+    This matches the user specified test spec with what tests Pants knows.
+
+    Each non fully qualified test spec may get matched with multiple targets.
+
+    :param possible_test_specs: an iterable of user specified test spec
+    :return: dict test_spec -> target
+    """
+    # dict of non fully qualified classname to a list of fully qualified test specs
+    non_fqcn_to_specs = defaultdict(list)
+    fqcn_to_specs = {}
+    for test_spec in self._test_to_target.keys():
+      fqcn = test_spec.classname
+      if fqcn in fqcn_to_specs:
+        raise TaskError('Dup found: {}'.format(fqcn))
```

Comment:
It looks like this fails a command line of `./pants test.junit --test=org.example.MyTestClass#testMethod1 --test=org.example.MyTestClass#testMethod2`. Do I have that right? If so, this seems like it should be allowed. As I read, it was allowed previously so this is a breaking change.

Reply:
Plan to merge this by EOD. @jsirois please let me know if there are additional concerns.

Reply:
It does appear to work. That said, I hope you find this as distrubing as I do:
https://github.com/pantsbuild/pants/blob/ffa655884c54a29afad154a90034cbe431c89480/src/python/pants/backend/jvm/tasks/junit_run.py#L244-L254

The `RegistryOfTests` constructed on line 254 would fail in the way I described if it had `match_test_spec` called on it. In other words this works as things happen to be used, but just looking at RegistryOfTests itself in isolation - it can be validly constructed with `{Test('f.q.c.n', 'method1'): ..., Test('f.q.c.n', 'method2'): ...}` - in which case it would be wonderful not to blow up here.

I still content the assertion is not necessary. I believe this change keeps your current functionality whil allowing for a RegistryOfTests initially constructed with >1 Test of the same classname:
```diff
diff --git a/src/python/pants/java/junit/junit_xml_parser.py b/src/python/pants/java/junit/junit_xml_parser.py
index 022d978c8..e9fbc662f 100644
--- a/src/python/pants/java/junit/junit_xml_parser.py
+++ b/src/python/pants/java/junit/junit_xml_parser.py
@@ -79,27 +79,19 @@ class RegistryOfTests(object):
     :return: dict test_spec -> target
     """
     # dict of non fully qualified classname to a list of fully qualified test specs
-    non_fqcn_to_specs = defaultdict(list)
-    fqcn_to_specs = {}
+    cn_to_specs = defaultdict(list)
     for test_spec in self._test_to_target.keys():
       fqcn = test_spec.classname
-      if fqcn in fqcn_to_specs:
-        raise TaskError('Dup found: {}'.format(fqcn))
-      fqcn_to_specs[fqcn] = test_spec
+      cn_to_specs[test_spec.classname].append(test_spec)
 
       non_fqcn = fqcn.split('.')[-1]
-      non_fqcn_to_specs[non_fqcn].append(test_spec)
+      cn_to_specs[non_fqcn].append(test_spec)
 
     matched_spec_to_target = {}
     unknown_tests = []
     for possible_test_spec in possible_test_specs:
-      # Match for fully qualified test spec
-      if possible_test_spec.classname in fqcn_to_specs:
-        matched_spec_to_target[possible_test_spec] = self._test_to_target[
-          fqcn_to_specs[possible_test_spec.classname]]
-      # Match for non fully qualified test specs
-      elif possible_test_spec.classname in non_fqcn_to_specs:
-        for full_spec in non_fqcn_to_specs[possible_test_spec.classname]:
+      if possible_test_spec.classname in cn_to_specs:
+        for full_spec in cn_to_specs[possible_test_spec.classname]:
           new_fully_qualified_test_spec = Test(full_spec.classname, possible_test_spec.methodname)
           matched_spec_to_target[new_fully_qualified_test_spec] = self._test_to_target[full_spec]
       else:
```

Reply:
Nice I like the simplified code!